docs/proposals/post-deployment-hooks.md

```diff
@@ -0,0 +1,187 @@
+  1. The pod will not enter a ready status until the hook has completed successfully.
+    1. Service endpoints will not be created for the pod until the pod has a ready status.
+  1. If the hook fails, the pod's creation is considered a failure, and the retry behavior is restart-policy driven in the usual way.
+
+Because deployments are represented as replication controllers, lifecycle hooks defined for containers are executed for every container in the replica set for the deployment. This results in disadvantages when applied to certain use cases:
+
+1. The hooks for all pods in the deployment will race, making the hooks extremely difficult to make correct (e.g., the hooks would generally need to be reentrant, idempotent, and safely concurrent).
+  1. Hooks which attempt to mutate shared state (e.g. a database) are provided with no support for coordination.
+
+##### Downstream post-deployment hooks
+
+Post-deployment hooks implemented external to the Kubelet provide a different set of guarantees:
+
+1. Hooks can be bound to the logical deployment lifecycle, enabling hook executions decoupled from replication mechanics.
+  1. For example, races can be avoided by defining a hook that executes exactly once per deployment regardless of the replica count.
+2. Hooks defined in terms of deployments are conceptually easier to reason about from the perspective of a user defining a deployment workflow.
+
+There are special considerations imposed by such an external hook mechanism:
+
+1. The hooks are executed asynchronous relative to the pod lifecycle.
+2. The hook outcomes have no direct impact on pod lifecycle.
+
+These considerations introduce disadvantages in some use cases:
+
+1. Nothing prevents external connectivity races: when a pod's containers are created, they become routable to services and other pods within an application the moment the containers enter a ready state, likely before or during hook execution.
+2. Hook execution can't be atomically linked to the pod's status: there will be a window where the pod is in a ready state, and a failed hook execution must revert the pod's state back to failed when it would be preferable to have never allowed to the pod to become ready prior to hook success.
+
+## Use cases
+
+1. As a MySQL user, I want my application deployment to seed the database with data the first time the database is used.
+2. As a MySQL user, I want slave instance deployments to be automatically registered with the master.
+3. As a MongoDB user, I want database peer instances to be automatically clustered during deployments.
+4. As a Rails application author, I want to perform a Rails database migration following an application deployment.
+
+#### Use-case: Seeding a MySQL database
+
+Databases are often initialized with example or other data the first time they are deployed. Users should be able to express a hook with causes the database to be initialized automatically.
```

Comment:
I updated this based on other feedback, not sure if it's clear enough for you yet.